src/legacy/shrink.rs

```diff
@@ -0,0 +1,426 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Read};
+
+use thiserror::Error;
+
+use super::bitstream::{lsb, BitStream};
+
+const MIN_CODE_SIZE: u8 = 9;
+const MAX_CODE_SIZE: u8 = 13;
+
+const MAX_CODE: usize = (1 << MAX_CODE_SIZE) - 1;
+const INVALID_CODE: u16 = u16::MAX;
+const CONTROL_CODE: usize = 256;
+const INC_CODE_SIZE: u64 = 1;
+const PARTIAL_CLEAR: u64 = 2;
+
+// const HASH_BITS: usize = MAX_CODE_SIZE + 1; /* For a load factor of 0.5. */
+// const HASHTAB_SIZE: usize = 1 << HASH_BITS;
+const UNKNOWN_LEN: u16 = u16::MAX;
+
+#[derive(Error, Debug)]
+enum ShrinkError {
+    #[error("self-referential code")]
+    InvalidPrefixCode,
+
+    #[error("first code needs to be literal")]
+    FirstCodeNeedsToBeLiteral,
+
+    #[error("invalid code")]
+    InvalidCode,
+
+    #[error("prev code no longer valid")]
+    PrevCodeNoLongerValid,
+}
+struct CodeQueue {
+    next_idx: usize,
+    codes: [u16; MAX_CODE as usize - CONTROL_CODE + 1],
+}
+
+impl CodeQueue {
+    fn new() -> Self {
+        let mut codes = [0; MAX_CODE as usize - CONTROL_CODE + 1];
+        for (i, code) in (CONTROL_CODE as u16 + 1..=MAX_CODE as u16).enumerate() {
+            codes[i] = code;
+        }
+        Self { next_idx: 0, codes }
+    }
+
+    // Return the next code in the queue, or INVALID_CODE if the queue is empty.
+    fn next(&self) -> u16 {
+        //   assert(q->next_idx < sizeof(q->codes) / sizeof(q->codes[0]));
+        self.codes[self.next_idx]
+    }
+
+    /// Return and remove the next code from the queue, or return INVALID_CODE if
+    /// the queue is empty.
+    fn remove_next(&mut self) -> u16 {
+        let code = self.next();
+        if code != INVALID_CODE {
+            self.next_idx += 1;
+        }
+        code
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Codetab {
+    prefix_code: u16, // INVALID_CODE means the entry is invalid.
+    ext_byte: u8,
+    len: u16,
+    last_dst_pos: usize,
+}
+
+impl Default for Codetab {
+    fn default() -> Self {
+        Self {
+            prefix_code: INVALID_CODE,
+            ext_byte: 0,
+            len: 0,
+            last_dst_pos: 0,
+        }
+    }
+}
+
+impl Codetab {
+    pub fn new() -> [Self; MAX_CODE + 1] {
+        let mut codetab = [Codetab::default(); MAX_CODE + 1];
+        // Codes for literal bytes. Set a phony prefix_code so they're valid.
+        for i in 0..=u8::MAX as usize {
+            codetab[i].prefix_code = i as u16;
+            codetab[i].ext_byte = i as u8;
+            codetab[i].len = 1;
+        }
+
+        codetab
```

Comment:
More idiomatic in Rust and more likely to avoid initializing elements that are immediately overwritten:
```suggestion
        let mut codetab = 0..=u8::MAX
            .map(|i| Codetab {
                prefix_code: i as u16,
                ext_byte: i,
                len: 1,
            })
            .collect::Vec<_>();
        codetab.resize(MAX_CODE + 1, Codetab::default());
        codetab.try_into().unwrap()
```

Reply:
This still isn't addressed.

Reply:
implemented.